blocks/log.py

```diff
@@ -1,274 +1,418 @@
+class TrainingLog(Mapping):
+    r"""The log of training progress.
 
-    To support various backends (such as database tables) the
-    descendants of this class are expected to override the
-    `__getattr__`, `__setattr__`, `__hasattr__` methods.
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Information is represented as nested dictionary
+    of iteration, key values.
 
-    Attributes
-    ----------
-    iterations_done : int
-        The number of iterations done.
-    epochs_done : int
-        The number of epochs done.
-    _epoch_ends : list
-        The numbers of the epochs last iterations.
+    In addition to the set of records of the training progress, a training
+    log has a status object whose attributes describe the current state of
+    training e.g. whether training has started, whether termination has
+    been requested, etc.
 
-    .. todo::
+    Parameters
+    ----------
+    hash_ : str, optional
+        A 12-byte hexidecimal string that identifies this experiment. If
+        given and a database backend is used, it will try to reload the
+        log. If not, a new hash will be created for this log.
+    backend : {'default', 'mongodb', 'sqlite'}, optional
+        The log can be stored in a variety of backends. The `default`
+        backend stores the values in a nested dictionary (a
+        :class:`DefaultOrderedDict` to be precise). This is a simple and
+        fast option, but the log is difficult to access for analysis during
+        training, and can only be saved as a pickled Python object.
+
+        The `mongo` option uses a MongoDB database server to store the
+        logs. The `sqlite` value will use a local SQLite file to store the
+        results. Both of these backends will allow you to analyze results
+        easily even during training, and will allow you to save multiple
+        experiments to the same database.
+    \*\*kwargs
+        Additional keyword arguments are passed to the log's backend e.g.
+        database configuration. See the relevant classes for details.
 
-        We need some notion of an attributes property. Examples of possible
-        properties include a docstring, a priority level to be used by a
-        printing extension to decide whether a certain attribute should be
-        printed or not.
+    Attributes
+    ----------
+    info : dict
+        A dictionary with information (metadata) about this experiment,
+        such as a description of the model, the package versions used, etc.
+        It also contains the unique hash that identifies experiments (e.g.
+        in the database).
+    status : :class:`TrainingStatus`
+        This dictionary contains information on the current status of the
+        training e.g. how many iterations/epochs have been completed,
+        whether training has finished, if errors occurred, etc.
+    hidden : dict
```

Comment:
Well, log[0].cost is (log[0]).cost

Why not use the fact that setitem can take multiple parameters? Then you 
couild write log[0,priority].cost =4 where behind the scenes 
log[0,priority] would return you a view of log, linked to the log, whose 
setattr sets the priority to 4. However, you cannot put something into 
the log with multiple priorities, e.g the following must fail:

log[0,log.INFO].cost =10
log[0,log.DEBUG].cost = 11 #exception raised

also, log[0] is equivalent to log[0,some_default_level]

Since we disallow having the same value with different priorities, 
getattr is easy - it just doesn't look at the priorities, hence the 
following might be the syntax:

log[0,log.INFO].cost =10

# log[0,log.DEBUG].cost = 11 #exception raised

assert log[0, log.DEBUG].cost == 10

#and also
  assert log[0].cost == 10

On 3/24/2015 2:08 PM, Bart van Merriënboer wrote:
>
> In blocks/log.py 
> <https://github.com/bartvm/blocks/pull/476#discussion_r27025057>:
>
> > -        We need some notion of an attributes property. Examples of possible
> > -        properties include a docstring, a priority level to be used by a
> > -        printing extension to decide whether a certain attribute should be
> > -        printed or not.
> > +    Attributes
> > +    ----------
> > +    info : dict
> > +        A dictionary with information (metadata) about this experiment,
> > +        such as a description of the model, the package versions used, etc.
> > +        It also contains the unique hash that identifies experiments (e.g.
> > +        in the database).
> > +    status : :class:`TrainingStatus`
> > +        This dictionary contains information on the current status of the
> > +        training e.g. how many iterations/epochs have been completed,
> > +        whether training has finished, if errors occurred, etc.
> > +    hidden : dict
>
> Annotating log entries with a certain level makes sense, since that 
> separates the annotation from the interpretation as you mentioned.
>
> Implementing this requires a complete overhaul of the interface 
> though. You can't just do |log[0].cost = 4| or |log[0]['cost'] = 4| 
> anymore, because you need a way to annotate it somehow. The |logging| 
> module-way of doing it would be |log.info(0, 'cost', 4)| I guess, 
> which we could consider. I do think that it takes away from the 
> simplicity of just having a nested dictionary as the interface though. 
> If I log this way, it's not directly clear at all how I would go about 
> extracting data from the log. The |logging| module doesn't have this 
> problem, because it just writes everything to a file or screen, but 
> for us this accessibility is very important i.e. not being able to do 
> |[t, v['cost'] for t, v in log.items() if 'cost' in v]| is a big downside.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/bartvm/blocks/pull/476/files#r27025057>.
>